src/Npgsql/NpgsqlParameter.cs

```diff
@@ -381,6 +402,13 @@ public string? DataTypeName
     {
         get
         {
+            if (_dbType is { } dbType && TypeInfo?.Options.DbTypeResolver is { } dbTypeResolver)
```

Comment:
This check should be below the other cases. The getters should always return their own explicitly set values first.

When it comes to the order we should ideally check from most specific to least (datatypename > npgsqldbtype > dbtype). I'll have to see how breaking it is if we flip the order we have today in ResolveTypeInfo (where npgsqldbtype > datatypename)